apps/src/weblab/WebLab.js

```diff
@@ -0,0 +1,213 @@
+ */
+WebLab.prototype.init = function (config) {
+  if (!this.studioApp_) {
+    throw new Error("WebLab requires a StudioApp");
+  }
+
+  this.skin = config.skin;
+  this.level = config.level;
+
+  this.brambleHost = null;
+
+  if (this.level.lastAttempt) {
+    this.startSources = this.level.lastAttempt;
+  } else if (this.level.startSources && this.level.startSources.length > 0) {
+    try {
+      this.startSources = JSON.parse(this.level.startSources);
+    } catch (err) {
+      console.error("Unable to parse startSources list", err);
+    }
+  }
+
+  config.usesAssets = true;
+
+  config.makeYourOwn = false;
+  config.centerEmbedded = false;
+  config.wireframeShare = true;
+  config.noHowItWorks = true;
+
+  config.getCodeAsync = this.getCodeAsync.bind(this);
+
+  // Provide a way for us to have top pane instructions disabled by default, but
+  // able to turn them on.
+  config.showInstructionsInTopPane = true;
+  config.noInstructionsWhenCollapsed = true;
+
+  let onMount = function () {
+    this.setupReduxSubscribers(this.studioApp_.reduxStore);
+
+    // TODO: understand if we need to call studioApp
+    // Other apps call studioApp.init(). That sets up UI that is not present Web Lab (run, show code, etc) and blows up
+    // if we call it. It's not clear there's anything in there we need, although we may discover there is and need to refactor it
+    // this.studioApp_.init(config);
+  }.bind(this);
+
+  // Push initial level properties into the Redux store
+  this.studioApp_.setPageConstants(config, {
+    channelId: config.channel,
+    isProjectLevel: !!config.level.isProjectLevel,
+  });
+
+  function onUndo() {
+      this.brambleHost.undo();
+  }
+
+  function onRedo() {
+    this.brambleHost.redo();
+  }
+
+  function onShowPreview() {
+    this.brambleHost.hideTutorial();
+  }
+
+  function onShowTutorial() {
+    this.brambleHost.showTutorial();
+  }
+
+  let inspectorOn = false;
+  function onToggleInspector() {
+    inspectorOn = !inspectorOn;
+    if (inspectorOn) {
+      this.brambleHost.enableInspector();
+    } else {
+      this.brambleHost.disableInspector();
+    }
+  }
+
+  ReactDOM.render((
+    <Provider store={this.studioApp_.reduxStore}>
+      <WebLabView
+        onUndo={onUndo.bind(this)}
+        onRedo={onRedo.bind(this)}
+        onShowPreview={onShowPreview.bind(this)}
+        onShowTutorial={onShowTutorial.bind(this)}
+        onToggleInspector={onToggleInspector.bind(this)}
+        onMount={onMount}
+      />
+    </Provider>
+  ), document.getElementById(config.containerId));
+
+  this.studioApp_.notifyInitialRenderComplete(config);
+};
+
+WebLab.prototype.getCodeAsync = function () {
+  return new Promise (function (resolve,reject) {
+    if (this.brambleHost !== null) {
+      this.brambleHost.getBrambleCode(function (code) {
+        resolve(code);
+      });
+    } else {
+      // Bramble not installed yet - we have no code to return
+      resolve("");
```

Comment:
Are we expected to hit this? Is resolving to empty string the right thing to do, vs. rejecting?